### ISSUE TITLE
rdf:JSON?

### ISSUE THREAD

**Issue by kasei** [awslabs/SPARQL-CDTs#3]
Do we need to think about interop with the `rdf:JSON` type? At the very least, we might want a casting function similar to the `xsd` functions that can turn an `rdf:JSON` literal into a CDT literal.

**Comment by hartig** [comment]
> Do we need to think about interop with the `rdf:JSON` type?

Yes, we need to think about (and document!) how exactly the `rdf:JSON` datatype is related to the two datatypes in our spec. I would expect that most well-formed `rdf:JSON` literals that represent a JSON object are also well-formed `cdt:Map` literals; there may be a few exceptions similar to the exceptions that we discuss in our [Sec.4.4 Relationship to JSON](https://awslabs.github.io/SPARQL-CDTs/spec/latest.html#map-datatype-versus-json).

> At the very least, we might want a casting function similar to the `xsd` functions that can turn an `rdf:JSON` literal into a CDT literal.

Yes, absolutely!

I suggest we keep this issue on the backlog for the moment. These are things that can still be added after going public with the proposal.

**Comment by fsasaki** [comment]
The approach of the XSLT 3.0 specification may be an inspiration for this topic. It defines two functions for the conversion 
https://www.w3.org/TR/xslt-30/#func-json-to-xml 
https://www.w3.org/TR/xslt-30/#func-xml-to-json
esp. the options of "json to xml", except "validate", may be something to look at then designing a "json to map" implementation.

**Comment by gkellogg** [comment]
This relates to w3c/json-ld-syntax#425 where there is a desire to be able to use a datatype other than `rdf:JSON` on a JSON literal; in that case, it would be `geo:geoJSONLiteral`. This makes a further case for such a feature in JSON-LD, which could allow literals to come out with `cdt:List` or `cdt:Map` natively. Of course, JSON can't directly represent IRIs and datatyped literals without the use of JSON-LD mechanisms, and I think interpreting literal JSON-LD in a way that would allow it to be transformed using these capabilities would be out of scope. But, many use cases would be well served using raw JSON strings, numbers, and boolean values in a JSON literal.

**Comment by hartig** [comment]
Thanks @gkellogg! However, I don't quite get the intention of your comment. Is this merely a FYI, highlighting the relationship to that JSON-LD issue, or is there something specific that you want us to do here?

**Comment by gkellogg** [comment]
It’s more to show what natively representing literals in JSON-LD using data types other than rdf:JSON is generally useful. But, I would say that use cases for CDT literals that are restricted to JSON, and don’t depend on embedded Turtle-like syntax, is important.

**Comment by hartig** [comment]
Thanks @fsasaki. Yet, I think it can be much easier in our case here, because there is no need to translate into a completely different syntax.

For rdf:JSON to cdt:Map, I assume it is essentially the identity mapping. The only thing we need to check is whether there are special cases in the lexical space of rdf:JSON that are not directly well-formed lexical forms for cdt:Map.

As for cdt:Map to rdf:JSON, we may define the conversion function as a partial function that produces an error for every CDT literal that contains elements that are not possible as elements in JSON (e.g., IRIs, literals with unsupported datatypes).
